File: Data Structures & Algorithms/longest-palindromic-substring/submission-2.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.size();
        if (n == 0) return "";
        if (n == 1) {
            return s;
        }
        int maxLength = -1;
        string maxString = "";
        int left, right;

        for (int i = 0; i < n; i++) {
            //for odd length palindromes
            left = i, right = i;
            while (left >= 0 && right < n) {
                if (s[left] != s[right]) {
                    break;
                }
                left--;
                right++;
            }
            ++left;
            --right;
            if ((right - left + 1) > maxLength) {
                maxLength = right - left + 1;
                maxString = s.substr(left, right - left + 1);
            }

            //for even length palindromes
            left = i, right = i + 1;

            while (left >= 0 && right < n) {
                if (s[left] != s[right]) {
                    break;
                }
                left--;
                right++;
            }
            ++left;
            --right;
            if ((right - left + 1) > maxLength) {
                maxLength = right - left + 1;
                maxString = s.substr(left, right - left + 1);
            }
        }
        
        return maxString;
    }
};

File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> freq = new HashMap<>();
        int left = 0, maxLength = 0;

        for (int right = 0; right < s.length(); right++) {
            freq.put(s.charAt(right), freq.getOrDefault(s.charAt(right), 0) + 1);

            while (freq.getOrDefault(s.charAt(right), 0) > 1) {
                freq.put(s.charAt(left), freq.getOrDefault(s.charAt(left), 0) - 1);
                ++left;
            }

            maxLength = Math.max(maxLength, right - left + 1);
        }
        
        return maxLength;
    }
}
